src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,635 @@
+/*
+ * Copyright (c) 2023 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <openssl/opensslv.h>
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+
+#include "soter/soter_rsa_key.h"
+
+#include <string.h>
+
+#include <openssl/bn.h>
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/param_build.h>
+#include <openssl/rsa.h>
+
+#include "soter/openssl/soter_rsa_key_utils.h"
+#include "soter/soter_portable_endian.h"
+#include "soter/soter_wipe.h"
+
+soter_status_t soter_engine_specific_to_rsa_pub_key(const soter_engine_specific_rsa_key_t* engine_key,
+                                                    soter_container_hdr_t* key,
+                                                    size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    int rsa_mod_size;
+    size_t output_length;
+    uint32_t* pub_exp;
+    // This bigint is reused multiple times: read into it, serialize it, read into it...
+    BIGNUM* tmp = NULL;
+
+    if (!key_length) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (EVP_PKEY_RSA != EVP_PKEY_id(pkey)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    // Relevant pkey params
+    // OSSL_PKEY_PARAM_BITS, "bits", int
+    // OSSL_PKEY_PARAM_RSA_N, "n", big uint
+    // OSSL_PKEY_PARAM_RSA_E, "e", big uint
+
+    // See https://docs.cossacklabs.com/themis/spec/asymmetric-keypairs/rsa/ for key layout
+
+    if (!EVP_PKEY_get_int_param(pkey, OSSL_PKEY_PARAM_BITS, &rsa_mod_size)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+    rsa_mod_size = (rsa_mod_size + 7) / 8;
+
+    if (!is_mod_size_supported(rsa_mod_size)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    output_length = rsa_pub_key_size(rsa_mod_size);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    if (!EVP_PKEY_get_bn_param(pkey, OSSL_PKEY_PARAM_RSA_N, &tmp)) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    if (BN_bn2binpad(tmp, (unsigned char*)(key + 1), rsa_mod_size) == -1) {
```

Comment:
should we validate `key` that it is not null before accessing? above we checked only key_length

Reply:
But we have a check above, lines 74..78
```
    if ((!key) || (output_length > *key_length)) {
        *key_length = output_length;
        res = SOTER_BUFFER_TOO_SMALL;
        goto err;
    }
```